File: symfony/src/Service/Content/DraftSessionService.php
<?php

namespace App\Service\Content;

use App\Entity\Content\DraftSession;
use App\Repository\Content\DraftSessionRepository;
use Symfony\Component\HttpKernel\Exception\ConflictHttpException;
use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
use Symfony\Component\HttpKernel\Exception\UnprocessableEntityHttpException;

class DraftSessionService
{
    public function __construct(
        private DraftSessionRepository $repository,
    ) {}

    public function create(array $data): DraftSession
    {
        throw new \LogicException('Not implemented');
    }

    public function update(DraftSession $entity, array $data): DraftSession
    {
        throw new \LogicException('Not implemented');
    }

    public function start(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('DraftSession not found: ' . $id);
        $entity->start();
        $this->repository->save($entity, flush: true);
    }

    public function abandon(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('DraftSession not found: ' . $id);
        $entity->abandon();
        $this->repository->save($entity, flush: true);
    }

    public function complete(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('DraftSession not found: ' . $id);
        $entity->complete();
        $this->repository->save($entity, flush: true);
    }

    public function isFull(int $id): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('DraftSession not found: ' . $id);
        $result = $entity->isFull();
        $this->repository->save($entity, flush: true);
        return $result;
    }
    public function transitionWaitingForPlayersToDrafting(int $id): object
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \Symfony\Component\HttpKernel\Exception\NotFoundHttpException();

        $entity->assertTransition($entity->getStatus(), 'Drafting');

        $entity->setStatus('Drafting');
        $entity->start(); // @after

        $this->repository->save($entity, flush: true);
        return $entity;
    }

    public function transitionDraftingToCompleted(int $id): object
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \Symfony\Component\HttpKernel\Exception\NotFoundHttpException();

        $entity->assertTransition($entity->getStatus(), 'Completed');

        $entity->setStatus('Completed');
        $entity->complete(); // @after

        $this->repository->save($entity, flush: true);
        return $entity;
    }

    #[\Symfony\Component\Security\Http\Attribute\IsGranted(['ROLE_ADMIN'])]
    public function transitionDraftingToAbandoned(int $id): object
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \Symfony\Component\HttpKernel\Exception\NotFoundHttpException();

        $entity->assertTransition($entity->getStatus(), 'Abandoned');

        $entity->setStatus('Abandoned');
        $entity->abandon(); // @after

        $this->repository->save($entity, flush: true);
        return $entity;
    }

    #[\Symfony\Component\Security\Http\Attribute\IsGranted(['ROLE_ADMIN'])]
    public function transitionWaitingForPlayersToAbandoned(int $id): object
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \Symfony\Component\HttpKernel\Exception\NotFoundHttpException();

        $entity->assertTransition($entity->getStatus(), 'Abandoned');

        $entity->setStatus('Abandoned');
        $entity->abandon(); // @after

        $this->repository->save($entity, flush: true);
        return $entity;
    }

    public function transitionCompletedToDrafting(int $id): never
    {
        throw new \Symfony\Component\HttpKernel\Exception\ConflictHttpException('Transition Completed -> Drafting is not allowed');
    }

    public function transitionAbandonedToDrafting(int $id): never
    {
        throw new \Symfony\Component\HttpKernel\Exception\ConflictHttpException('Transition Abandoned -> Drafting is not allowed');
    }
}

File: spring/src/main/java/cardsproject/service/content/DraftSessionService.java
package cardsproject.service.content;

import cardsproject.domain.content.DraftSession;
import cardsproject.repository.content.DraftSessionRepository;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Optional;
import cardsproject.domain.content.DraftSessionStatusType;
import cardsproject.domain.content.DraftSessionDraftTypeType;

@Service
public class DraftSessionService {

    private final DraftSessionRepository repository;

    public DraftSessionService(DraftSessionRepository repository) {
        this.repository = repository;
    }

    public List<DraftSession> findAll() {
        return repository.findAll();
    }

    public Optional<DraftSession> findById(Long id) {
        return repository.findById(id);
    }

    public DraftSession save(DraftSession entity) {
        validate(entity);
        return repository.save(entity);
    }

    public void delete(Long id) {
        repository.deleteById(id);
    }

    public void applyPatch(DraftSession entity, java.util.Map<String, Object> patch) {
        if (patch.containsKey("status")) entity.setStatus(DraftSessionStatusType.valueOf(patch.get("status").toString()));
        if (patch.containsKey("draftType")) entity.setDraftType(DraftSessionDraftTypeType.valueOf(patch.get("draftType").toString()));
        if (patch.containsKey("seats") && patch.get("seats") != null) entity.setSeats(Integer.valueOf(patch.get("seats").toString()));
        if (patch.containsKey("timePerPickSeconds") && patch.get("timePerPickSeconds") != null) entity.setTimePerPickSeconds(Integer.valueOf(patch.get("timePerPickSeconds").toString()));
        if (patch.containsKey("createdAt") && patch.get("createdAt") != null) entity.setCreatedAt(java.time.LocalDateTime.parse(patch.get("createdAt").toString()));
        if (patch.containsKey("completedAt") && patch.get("completedAt") != null) entity.setCompletedAt(java.time.LocalDateTime.parse(patch.get("completedAt").toString()));
        if (patch.containsKey("cardSetId") && patch.get("cardSetId") != null) entity.setCardSetId(Long.valueOf(patch.get("cardSetId").toString()));
    }
    private void validate(DraftSession entity) {
        if (entity.getCompletedAt() != null && !(DraftSessionStatusType.COMPLETED.equals(entity.getStatus()))) throw new IllegalStateException("completed_at can only be set when draft status is Completed");
    }

    public void start(Long id) {
        DraftSession entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("DraftSession not found: " + id));
        entity.start();
        repository.save(entity);
    }

    public void abandon(Long id) {
        DraftSession entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("DraftSession not found: " + id));
        entity.abandon();
        repository.save(entity);
    }

    public void complete(Long id) {
        DraftSession entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("DraftSession not found: " + id));
        entity.complete();
        repository.save(entity);
    }

    public Boolean isFull(Long id) {
        DraftSession entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("DraftSession not found: " + id));
        Boolean result = entity.isFull();
        repository.save(entity);
        return result;
    }

    public DraftSession transitionWaitingForPlayersToDrafting(Long id) {
        DraftSession entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("DraftSession not found: " + id));
        entity.assertTransition(DraftSessionStatusType.DRAFTING);
        entity.setStatus(DraftSessionStatusType.DRAFTING);
        entity.start(); // @after
        return repository.save(entity);
    }

    public DraftSession transitionDraftingToCompleted(Long id) {
        DraftSession entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("DraftSession not found: " + id));
        entity.assertTransition(DraftSessionStatusType.COMPLETED);
        entity.setStatus(DraftSessionStatusType.COMPLETED);
        entity.complete(); // @after
        return repository.save(entity);
    }

    public DraftSession transitionDraftingToAbandoned(Long id) {
        DraftSession entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("DraftSession not found: " + id));
        entity.assertTransition(DraftSessionStatusType.ABANDONED);
        entity.setStatus(DraftSessionStatusType.ABANDONED);
        entity.abandon(); // @after
        return repository.save(entity);
    }

    public DraftSession transitionWaitingForPlayersToAbandoned(Long id) {
        DraftSession entity = repository.findById(id)
            .orElseThrow(() -> new RuntimeException("DraftSession not found: " + id));
        entity.assertTransition(DraftSessionStatusType.ABANDONED);
        entity.setStatus(DraftSessionStatusType.ABANDONED);
        entity.abandon(); // @after
        return repository.save(entity);
    }

    public void transitionCompletedToDrafting(Long id) {
        throw new IllegalStateException("Transition Completed -> Drafting is not allowed");
    }

    public void transitionAbandonedToDrafting(Long id) {
        throw new IllegalStateException("Transition Abandoned -> Drafting is not allowed");
    }
}
